docs/notebooks/asynchronous_bayesian_optimization.pct.py

```diff
@@ -0,0 +1,373 @@
+# Next we define two helper functions, one is to create a certain number of worker processes, and another is to terminate them once we are done.
+
+# %%
+
+def create_worker_processes(n_workers, points_queue, obseverations_queue):
+    observer_processes = []
+    for i in range(n_workers):
+        worker_proc = Process(target=observer_proc, args=(points_queue, obseverations_queue))
+        worker_proc.daemon = True
+        worker_proc.start()
+
+        observer_processes.append(worker_proc)
+
+    return observer_processes
+
+def terminate_processes(processes):
+    for prc in processes:
+        prc.terminate()
+        prc.join()
+        prc.close()
+
+# %% [markdown]
+# Finally we set some common parameters. See comments below for explanation of what each one means.
+# %%
+# Number of worker processes to run simultaneously
+# Setting this to 1 will turn both setups into non-batch sequential optimization
+num_workers = 3
+# Number of iterations to run the sycnhronous scenario for
+num_iterations = 10
+# Number of observations to collect in asynchronous scenario
+num_observations = num_workers * num_iterations
+# Set this flag to False to disable sleep delays in case you want the notebook to execute quickly
+enable_sleep_delays = True
+
+# %% [markdown]
+# ## Asynchronous optimization
+# This section runs the asynchronous optimization routine. We first setup the [ask/tell optimizer](ask_tell_optimization.ipynb). Notice that we use **AsyncEfficientGlobalOptimization** rule specifically designed for asynchronous scenarios. Next we create thread-safe queues for points and observations, and run the optimization loop.
+#
+# Crucially, even though we are using batch acquisition function Local Penalization, we specify batch size of 1. This is because we don't really want a batch. Since the amout of workers we have is fixed, whenever we see a new observation we only need one point back. However this process can only be done with acquisition funcitons that implement greedy batch collection strategies, because they are able to take into account points that are currently being observed (in Trieste we call them "pending"). Trieste currently provides two such functions: Local Penalization and GIBBON.
+
+# %%
+
+# setup Ask Tell BO
+model_spec = build_model(initial_data)
+model = create_model(model_spec)
+
+local_penalization_acq = LocalPenalizationAcquisitionFunction(search_space, num_samples=2000)
+local_penalization_rule = AsyncEfficientGlobalOptimization(num_query_points=1, builder=local_penalization_acq)  # type: ignore
+
+async_bo = AskTellOptimizer(search_space, initial_data, model, local_penalization_rule)
+
+# retrieve process id for nice logging
+pid = os.getpid()
+# create point and observation queues
+m = Manager()
+pq = m.Queue()
+oq = m.Queue()
+# keep track of all workers we have launched
+observer_processes = []
+# counter to keep track of collected observations
+points_observed = 0
+
+start = timeit.default_timer()
+try:
```

Comment:
a bit of education for me - can you clarify why is `try` needed here?

Reply:
to clean up running processes in case anything goes wrong and the code falls with exception. Note that we are not actually using `except`, only `try/finally`